File: app/src/main/java/com/example/testapp/andserver/ServerHost.java
package com.example.testapp.andserver;

import android.util.Log;

import com.example.testapp.unit.HttpUnit;
import com.yanzhenjie.andserver.AndServer;
import com.yanzhenjie.andserver.Server;

import java.net.InetAddress;
import java.util.concurrent.TimeUnit;

public class ServerHost {

    private Server mServer;
    private String TAG = "ServerHost";


    public ServerHost(InetAddress inetAddress) {

        mServer = AndServer.serverBuilder()
                .inetAddress(inetAddress)
                .port(8088)
                .timeout(10, TimeUnit.SECONDS)
                .listener(new ServerListener())
                .build();
    }

    public void startServer() {
        if (!mServer.isRunning()) {
            mServer.startup();
        }
    }

    public void stopServer() {
        if (mServer.isRunning()) {
            mServer.shutdown();
        } else {
            Log.w("AndServer", "The server has not started yet.");
        }
    }


    private class ServerListener implements Server.ServerListener {

        @Override
        public void onStarted() {
            Log.i(TAG, "onStarted");

            InetAddress address = HttpUnit.getLocalIPAddress();
            Log.i(TAG, mServer.getInetAddress().getHostAddress());
            Log.i(TAG, "Host : " + address.getHostAddress());
        }

        @Override
        public void onStopped() {
            Log.i(TAG, "onStopped");
        }

        @Override
        public void onException(Exception e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/com/example/testapp/andserver/ServerHost.kt
package com.example.testapp.andserver

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import com.anmnight.commlibrary.unit.HttpUnit
import com.example.testapp.andserver.manager.ServerHostBroadcastManager

import com.yanzhenjie.andserver.AndServer
import com.yanzhenjie.andserver.Server
import java.util.concurrent.TimeUnit

class ServerHost : Service() {


    private var mServer: Server? = null
    private val TAG = "ServerHost"


    override fun onCreate() {
        super.onCreate()

        if (mServer == null) {
            mServer = initServer()
        }

    }


    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        startServer()
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        stopServer()
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }


    private fun initServer(): Server {
        return AndServer.serverBuilder()
                .inetAddress(HttpUnit.getLocalIPAddress())
                .port(8088)
                .timeout(10, TimeUnit.SECONDS)
                .listener(ServerListener())
                .build()
    }

    private fun startServer() {

        if (!mServer!!.isRunning) {
            mServer?.startup()
        }


    }

    private fun stopServer() {
        if (mServer!!.isRunning) {
            mServer!!.shutdown()
        } else {
            Log.w("AndServer", "The server has not started yet.")
        }
    }

    inner class ServerListener : Server.ServerListener {

        override fun onStarted() {

            val address = "http://" + HttpUnit.getLocalIPAddress().hostAddress + ":8088"
            sendBroadcast(ServerHostBroadcastManager.startIntent(address))

        }

        override fun onStopped() {
            sendBroadcast(ServerHostBroadcastManager.stopIntent())
        }

        override fun onException(e: Exception) {
            e.printStackTrace()
        }
    }


}
